parlai/core/torch_classifier_agent.py

```diff
@@ -217,15 +220,31 @@ def _update_confusion_matrix(self, batch, predictions):
         """
         Update the confusion matrix given the batch and predictions.
 
-        :param batch:
-            a Batch object (defined in torch_agent.py)
         :param predictions:
             (list of string of length batchsize) label predicted by the
             classifier
+        :param batch:
+            a Batch object (defined in torch_agent.py)
         """
-        for i, pred in enumerate(predictions):
-            label = batch.labels[i]
-            self.metrics['confusion_matrix'][(label, pred)] += 1
+        f1_dict = {}
+        for class_name in self.class_list:
+            if not self.opt['get_all_metrics']:
+                class_name = self.ref_class
```

Comment:
the if doesn't belong here. `if not self.opt['get_all_metrics']` then you are computing the same metrics `len(self.class_list)` times.